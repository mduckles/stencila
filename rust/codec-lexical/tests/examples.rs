use std::{fs::read_to_string, path::PathBuf};

use codec::{
    common::{eyre::Result, glob::glob},
    format::Format,
    EncodeOptions,
};

use codec_lexical::{decode, encode};
use common_dev::insta::{assert_json_snapshot, assert_snapshot, assert_yaml_snapshot};

/// Decode each example of a Lexical document and create JSON and Lexical snapshots
/// including snapshots for losses
#[test]
fn examples() -> Result<()> {
    let pattern = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/examples")
        .canonicalize()?
        .to_string_lossy()
        .to_string()
        + "/**/*";

    for path in glob(&format!("{pattern}.lexical"))?
        .chain(glob(&format!("{pattern}.koenig"))?)
        .flatten()
    {
        let format = Format::from_path(&path);
        let ext = format.extension();

        let name = path
            .file_name()
            .map(|name| name.to_string_lossy())
            .and_then(|name| name.strip_suffix(&[".", &ext].concat()).map(String::from))
            .expect("should have format suffix");

        let original = read_to_string(path)?;

        let (article, info) = decode(&original, None)?;

        assert_json_snapshot!(format!("{name}.json"), article);
        assert_yaml_snapshot!(format!("{name}.decode.losses"), info.losses);

        let (lexical, info) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format.clone()),
                compact: Some(false),
                ..Default::default()
            }),
        )?;

        assert_snapshot!(format!("{name}.{ext}"), lexical);
        assert_yaml_snapshot!(format!("{name}.encode.losses"), info.losses);

        let (lexical, ..) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format),
                standalone: Some(true),
                compact: Some(false),
                ..Default::default()
            }),
        )?;
        assert_snapshot!(format!("{name}.standalone.{ext}"), lexical);
    }

    Ok(())
}
