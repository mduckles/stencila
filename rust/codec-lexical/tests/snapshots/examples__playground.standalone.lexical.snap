---
source: rust/codec-lexical/tests/examples.rs
expression: lexical
---
{
  "root": {
    "type": "root",
    "children": [
      {
        "type": "html",
        "html": "<stencila-heading id=hea_53NpLN7XA4P3ZaZDK6W1Jd depth=0 ancestors='' level=1><h1 slot=content><stencila-text id=txt_Mvuz5Kn5oJALW9fuuyTtiz depth=1 ancestors=Heading>Welcome to the playground</stencila-text></h1></stencila-heading><stencila-quote-block id=qtb_4bNYzJWoHjLKhDFwbmuV7t depth=0 ancestors=''><blockquote slot=content><stencila-paragraph id=pgh_Fa6T2ZfcrQ46pWGFmbjeF depth=1 ancestors=QuoteBlock><p slot=content><stencila-text id=txt_FXwfHkinw6REErYpVriid5 depth=2 ancestors=QuoteBlock.Paragraph>In case you were wondering what the black box at the bottom is â€“ it&#x27;s the debug view, showing the current state of the editor. You can disable it by pressing on the settings control in the bottom-left of your screen and toggling the debug view setting.</stencila-text></p></stencila-paragraph></blockquote></stencila-quote-block><stencila-paragraph id=pgh_DBSdKtZyPoQsEZsVTYUFa5 depth=0 ancestors=''><p slot=content><stencila-text id=txt_Szvd23Kwda4AxzhMwKpsec depth=1 ancestors=Paragraph>The playground is a demo environment built with </stencila-text><stencila-code-inline id=cdi_5ZmjMJQ8ixn26dQMqjgwDP depth=1 ancestors=Paragraph code=@lexical/react><code>@lexical&#x2F;react</code></stencila-code-inline><stencila-text id=txt_Kb2aVo5WzKsrsX14nBLTMV depth=1 ancestors=Paragraph>. Try typing in </stencila-text><stencila-strong id=stg_LYjZagdMYH21PS5HAqbCef depth=1 ancestors=Paragraph><strong><stencila-text id=txt_4wDRaZXtedAMpERS6s4aMz depth=2 ancestors=Paragraph.Strong>some text</stencila-text></strong></stencila-strong><stencila-text id=txt_EcpFR85vkV3LarwA7UrfpB depth=1 ancestors=Paragraph> with </stencila-text><stencila-emphasis id=emp_LHFH8pQuDxz71GmY97mUSw depth=1 ancestors=Paragraph><em><stencila-text id=txt_UA1pYz4SnSrAvE8VWBBNH9 depth=2 ancestors=Paragraph.Emphasis>different</stencila-text></em></stencila-emphasis><stencila-text id=txt_DzS87dps2okDPqaY6BgQSw depth=1 ancestors=Paragraph> formats.</stencila-text></p></stencila-paragraph><stencila-paragraph id=pgh_HwHr4Vhf4FZTawxxQacu1d depth=0 ancestors=''><p slot=content><stencila-text id=txt_9ADphS4ihmwzStPR531mpi depth=1 ancestors=Paragraph>Make sure to check out the various plugins in the toolbar. You can also use </stencila-text><stencila-text id=txt_2DMQDuM4Zt44GHtiQzcdYw depth=1 ancestors=Paragraph>LOST HashTag</stencila-text><stencila-text id=txt_3wLRoPzWoLaMWxhYPE7sV3 depth=1 ancestors=Paragraph> or @-mentions too!</stencila-text></p></stencila-paragraph><stencila-paragraph id=pgh_CgUXY1bdypH2vriW9AX6Xa depth=0 ancestors=''><p slot=content><stencila-text id=txt_M6EP6oZ7jonmEgsYt8opjv depth=1 ancestors=Paragraph>If you&#x27;d like to find out more about Lexical, you can:</stencila-text></p></stencila-paragraph><stencila-paragraph id=pgh_JH1XNG5MfR4PFWdTbKZTnz depth=0 ancestors=''><p slot=content><stencila-text id=txt_2uXjnX9vNsgDgwuCe4SM8j depth=1 ancestors=Paragraph>LOST List</stencila-text></p></stencila-paragraph><stencila-paragraph id=pgh_RY8mQMjT7mjkqRvWNT2ZwE depth=0 ancestors=''><p slot=content><stencila-text id=txt_LFL1hsX5V3p62zAjtKbwCF depth=1 ancestors=Paragraph>Lastly, we&#x27;re constantly adding cool new features to this playground. So make sure you check back here when you next get a chance ðŸ™‚.</stencila-text></p></stencila-paragraph>"
      }
    ]
  }
}
